pep-9999.rst

```diff
@@ -0,0 +1,217 @@
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file, under the ``[repositories]`` section::
+
+
+    [repositories]
+
+    [repostiories.pypi]
+    url = "https://pypi.org/simple/"
+
+    [repostiories.myrepo1]
+    url = "http://myserver2.org/simple"
+
+
+A repository has the following attributes::
+
+    ==================  ===========  =============================================
+    Attribute           Type         Description
+    ==================  ===========  =============================================
+    final               bool         Make it impossible to override/remove the repository configuration (default: false)
+    proxy_port          int          The port of the proxy server to use (default: 80)
+    proxy_url           str          The URL to the proxy server
+    shadow              bool         Use package from this index, even if it was found in a previous index (default: false)
+    url                 str          URL to repository
+    ==================  ===========  =============================================
+
+
+The auth.toml file
+''''''''''''''''''
+
+Credentials are stored in the ``auth.toml`` file, under the ``[authentication]`` section. There is a correlation between the repository name and the authentication section's name::
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+Instead of specifying the credentials in this file, Keyring can be used::
+
+    keyring set https://pypi.org/simple/ __token__
+    keyring set <repository URL> <private-repository username>
+
+
+A repository's credentials has the following attributes::
+
+    ==================  ===========  ===============================================
+    Attribute           Type         Description
+    ==================  ===========  ===============================================
+    cert                str          Path to certificate authority bundle (*.pem)
+    cert-client         str          Path to client certificate (*.pem)
+    password            str          The password for the repository, base64-encoded. When token is used, enter it here in clear-text.
```

Comment:
"Base 64" seems like an implementation concern. It should probably be plain text, and tools can convert when necessary, unless there's some concern here I haven't thought of (in which case it should be written in the doc - and being able to tell between a token and a password is something I thought of and it can't be solved with this rule ;) )

Reply:
So the reason for base64 is simply to avoid having people "look over the shoulder" and see the credentials that way.

But I guess you might be right about the token... this could complicate things.